File: cf_sdk/src/main/java/com/example/cf_sdk/logic/changebanksdk/SingleUseCase.java
package com.example.cf_sdk.logic.changebanksdk;

import com.example.cf_sdk.logic.changebanksdk.executor.ExecutionThread;
import com.example.cf_sdk.logic.changebanksdk.parameter.Parameters;
import com.example.cf_sdk.logic.changebanksdk.util.ChangebankError;

import io.reactivex.Single;
import io.reactivex.SingleObserver;

/**
 *
 * Use case to be used when there is only one result expected.
 */

public abstract class SingleUseCase<P extends Parameters, RESULT>
        extends UseCase<P, Single<RESULT>, SingleObserver<RESULT>> {
    private ExecutionThread mBackgroundExecutionThread;
    protected ExecutionThread mUiExecutionThread;
    protected ChangebankError mErrorHandler;

    protected SingleUseCase(ExecutionThread backgroundExecutionThread,
                      ExecutionThread uiExecutionThread,
                      ChangebankError errorHandler) {
        mBackgroundExecutionThread = backgroundExecutionThread;
        mUiExecutionThread = uiExecutionThread;
        mErrorHandler = errorHandler;
    }

    @Override
    protected abstract Single<RESULT> buildUseCaseObservable(P parameters);

    @SuppressWarnings("unchecked")
    @Override
    public void execute(SingleObserver<RESULT> observer, P parameters) {
        buildUseCaseObservable(parameters)
                .subscribeOn(mBackgroundExecutionThread.getScheduler())
                .observeOn(mUiExecutionThread.getScheduler())
                .compose(mErrorHandler)
                .subscribe(observer);
    }
}

File: cf_sdk/src/main/java/com/example/cf_sdk/changebankapi/usecase/SingleUseCase.kt
package com.example.cf_sdk.changebankapi.usecase

import com.example.cf_sdk.changebankapi.parameter.Parameters
import com.example.cf_sdk.changebankapi.usecase.ExecutionThread
import com.example.cf_sdk.changebankapi.usecase.UseCase
import com.example.cf_sdk.changebankapi.util.ChangebankError

import io.reactivex.Single
import io.reactivex.SingleObserver
import org.xml.sax.ErrorHandler
import retrofit2.Retrofit


/**
 *
 * Use case to be used when there is only one result expected.
 */
abstract class SingleUseCase<P : Parameters?, RESULT>(
    private val mBackgroundExecutionThread: ExecutionThread,
    protected var mUiExecutionThread: ExecutionThread,
    protected var mErrorHandler: ChangebankError<*>
) : UseCase<P, Single<RESULT>, SingleObserver<RESULT>>() {

    abstract override fun buildUseCaseObservable(parameters: P): Single<RESULT>

    override fun execute(observer: SingleObserver<RESULT>, parameters: P) {
        buildUseCaseObservable(parameters)
            .subscribeOn(mBackgroundExecutionThread.scheduler)
            .observeOn(mUiExecutionThread.scheduler)
            ?.subscribe(observer)
    }
}